### DISCUSSION TITLE
Call limit reached in hypotest fucntion

### DISCUSSION THREAD

**Discussion by ssaha1234** [scikit-hep/pyhf#1832] [Q&A] [answered]
Dear pyhf users, 
    I am running a hypotest with several signal hypotheses as shown below - 

```python
test_mus = np.linspace(0.2, 3, 60)
results = [
    pyhf.infer.hypotest(
        test_mu, data, model, return_expected_set=True, return_tail_probs=True
    )
    for test_mu in test_mus
]
```

Unfortunately, the Call Limit is being reached in the fit. I tried adding ` maxiter=1000000` inside the function, but I get an error saying this is not a valid argument.  Is there a way, I can add maximum iterations in the hypotest ? Let me know if you would need more details about the likelihood, or the errors. 
Thanks a lot !

Best, 
Shreya

**Comment by matthewfeickert** [comment]
@ssaha1234 `maxiter` is controlled by the optimizer (c.f. [docs for the different optimizers](https://pyhf.readthedocs.io/en/v0.6.3/_generated/pyhf.optimize.mixins.OptimizerMixin.html#optimizermixin)). So you'll need to set that attribute _before_ you perform inference.

Probably the most straightforward way to do this (in `pyhf` `v0.6.3`) is during the backend setup. Here's an example using NumPy and the SciPy optimizer

```python
import pyhf
print(f"pyhf version: v{pyhf.__version__}")  # pyhf version: v0.6.3
pyhf.set_backend("numpy", pyhf.optimize.scipy_optimizer(maxiter=2000000))
backend, optimizer = pyhf.get_backend()
print(f"{optimizer.maxiter=}")  # optimizer.maxiter=2000000
```

Though I'll note that if your fit is struggling here bumping the `maxiter` probably won't help, as iteration time isn't going to be the underlying problem.

**Comment by kratsg** [comment, marked as answer]
Here's a working example of what you intended to do. Note that a `hypotest` call will instantiate multiple optimizers (because it's not a single fit, but at least 5 separate fits using the specified optimizer).

```python
import pyhf

new_optimizer = pyhf.optimize.scipy_optimizer(maxiter=1000000)
pyhf.set_backend(pyhf.tensorlib, new_optimizer)

model = pyhf.simplemodels.uncorrelated_background([5], [10], [2.5])
data = [15] + model.config.auxdata
test_mu = 1.0

result = pyhf.infer.hypotest(
    test_mu, data, model, return_expected_set=True, return_tail_probs=True,
)
```

**Reply by ssaha1234** [reply]
Hi @matthewfeickert and @kratsg, 
    Thank you very much for the information and code snippets, I will implement it and check if I can make the fit work. 
@matthewfeickert I have seen a similar error before for a mle fit (but that is a single fit...)- https://github.com/scikit-hep/pyhf/discussions/1751, maxiter did solve the issue, which is why I thought it would work out here as well :)). I will try it out and let you know ! 
Thanks ! 
Best, 
Shreya

**Reply by ssaha1234** [reply]
Hi @matthewfeickert and @kratsg, 
   My fits worked when I increased the iterations ! Thanks a lot for all the help :))
Best, 
Shreya
